WIP Changes to allow quippy build with Python3.12

smjanke opened the pull request:
I changed the quippy Makefile to allow compilation with Python3.12, f90wrap==0.2.16 and meson build, which replaces the deprecated distutils. Additionally, I removed the requirement for of numpy<2 from quippy/setup.py, which does not seem to be necessary anymore.

The quippy compilation currently still fails, apparently because of a bug in likely f90wrap that causes duplicated type-declarations to appear in one of the f90wrap-converted files.

Please note that these changes do not allow compilation with Python 3.11 anymore. This is due to the changes to the quippy/Makefile in the quippy build instructions, line 118-119. These changes are necessary to allow compilation with meson, and incompatible with the previously used distutils.

smjanke committed:
Changed quippy Makefile for Python3.12

I changed the quippy Makefile to allow compilation with Python3.12
and f90wrap 0.2.16 and meson build, which replaces the
deprecated distutil.

smjanke committed:
Removed requirement numpy<2 from quippy/setup.py

bernstei commented:
> Please note that these changes do not allow compilation with Python 3.11 anymore.

Entirely failing to support python < 3.12 is not great.  Can we make this rule dependent on python version (e.g. parse the output of `${PYTHON} --version`) ?

jameskermode commented:
Yes, I won't merge this until we come up with a way to make it works with both old and new Python.

bernstei commented:
> Yes, I won't merge this until we come up with a way to make it works with both old and new Python.

Do you want me to try to work that out?  I can test this branch on our python 3.11 setup.

jameskermode commented:
That would be great. I'll try with Python 3.12 version and see if I can resolve the f90wrap problem.

bernstei commented:
I can no longer compile the main `public` branch

```
F90=gfortran LDFLAGS="" f2py-f90wrap --build-dir . -c -m _quippy f90wrap_Atoms_types.f90 f90wrap_Atoms.f90 f90wrap_System.f90 f90wrap_Dictionary.f90 f90wrap_DynamicalSystem.f90 f90wrap_nye_tensor.f90 f90wrap_Potential.f90 f90wrap_Potential_simple.f90 f90wrap_TB.f90 f90wrap_descriptors.f90 \
        -L. -lquip_nostub -L/home/cluster/bernstei/src/work/QUIP/source/QUIP_github/src/fox/objs.linux_x86_64_gfortran/lib -lFoX_sax -lFoX_wxml -lFoX_utils -lFoX_common -lFoX_fsys -L/opt/intel/oneapi/mkl/2022.1.0/lib/intel64 -lmkl_gf_lp64 -lmkl_sequential -lmkl_core -lpthread --fcompiler=gnu95 --f90flags="-ffree-line-length-none -fPIC"

!! f90wrap patched version of f2py - James Kermode <james.kermode@gmail.com> !!

Traceback (most recent call last):
  File "/home/cluster/bernstei/.local/bin/f2py-f90wrap", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/local/python/quippy/conda/2024_10_21/lib/python3.11/site-packages/f90wrap/scripts/f2py_f90wrap.py", line 219, in main
    numpy.f2py.main()
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/f2py/f2py2e.py", line 766, in main
    run_compile()
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/f2py/f2py2e.py", line 594, in run_compile
    build_backend = f2py_build_generator(backend_key)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/f2py/_backends/__init__.py", line 6, in f2py_build_generator
    from ._distutils import DistutilsBackend
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/f2py/_backends/_distutils.py", line 3, in <module>
    from numpy.distutils.core import setup, Extension
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/distutils/core.py", line 24, in <module>
    from numpy.distutils.command import config, config_compiler, \
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/distutils/command/config.py", line 19, in <module>
    from numpy.distutils.mingw32ccompiler import generate_manifest
  File "/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/distutils/mingw32ccompiler.py", line 27, in <module>
    from distutils.msvccompiler import get_build_version as get_build_msvc_version
ModuleNotFoundError: No module named 'distutils.msvccompiler'
make[1]: *** [Makefile:118: quippy/_quippy.cpython-311-x86_64-linux-gnu.so] Error 1
make[1]: Leaving directory '/home/cluster/bernstei/src/work/QUIP/source/QUIP_github/build/linux_x86_64_gfortran'
make: *** [Makefile:230: quippy] Error 2
```

bernstei commented:
I see this perhaps relevant thread https://github.com/numpy/numpy/issues/27405

bernstei commented:
When I use the workaround from that thread (still with regular `public` branch and python 3.11), I get a compilation error due to
```
./src.linux-x86_64-3.11/_quippymodule.c: In function ‘f2py_rout__quippy_f90wrap_system_module__s2a’:
./src.linux-x86_64-3.11/_quippymodule.c:46082:9: error: ‘ret_a_capi’ undeclared (first use in this function); did you mean ‘ret_a_Rank’?
46082 |     if (ret_a_capi != Py_None) slen(ret_a) = f2py_itemsize(ret_a);
      |         ^~~~~~~~~~
      |         ret_a_Rank
./src.linux-x86_64-3.11/_quippymodule.c:46082:9: note: each undeclared identifier is reported only once for each function it appears in
./src.linux-x86_64-3.11/_quippymodule.c: In function ‘f2py_rout__quippy_f90wrap_system_module__pad’:
./src.linux-x86_64-3.11/_quippymodule.c:46324:9: error: ‘ret_a_capi’ undeclared (first use in this function); did you mean ‘ret_a_Rank’?
46324 |     if (ret_a_capi != Py_None) slen(ret_a) = f2py_itemsize(ret_a);
      |         ^~~~~~~~~~
      |         ret_a_Rank
```
which eventually leads to 
```
error: Command "gcc -pthread -B /home/Software/python/miniconda3/compiler_compat -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /home/Software/python/miniconda3/include -fPIC -O2 -isystem /home/Software/python/miniconda3/include -fPIC -DNPY_DISABLE_OPTIMIZATION=1 -I./src.linux-x86_64-3.11/./src.linux-x86_64-3.11 -I/home/Software/python/miniconda3/extra/lib/python3.11/site-packages/numpy/core/include -I/home/Software/python/miniconda3/include/python3.11 -c ./src.linux-x86_64-3.11/_quippymodule.c -o ./src.linux-x86_64-3.11/_quippymodule.o -MMD -MF ./src.linux-x86_64-3.11/_quippymodule.o.d" failed with exit status 1
make[1]: *** [Makefile:118: quippy/_quippy.cpython-311-x86_64-linux-gnu.so] Error 1
make[1]: Leaving directory '/home/cluster/bernstei/src/work/QUIP/source/QUIP_github/build/linux_x86_64_gfortran'
make: *** [Makefile:230: quippy] Error 2
```

bernstei commented:
Note, also, that the current build system does not handle `PYTHON` being unset gracefully at all.

smjanke commented:
Very many apologies for this mess! This should not have been a merge request at this stage, but rather just give James the opportunity to look into an f90wrap compilation problem that appears for the Python3.12 compilation with f90wrap 0.2.15 and 0.2.16. Mostly, I was doing two things at once, and ended up thinking of a different, parallel project, which resulted in me doing a merge request instead of a pull request, which was very stupid.

Thank you very much for working on it! 
- Issue with incompatible build between Python 3.11 and 3.12: I was hoping to look into how to provide conditions that allows compilation with Python3.11 and Python 3.12 over the next few days. 

- Numpy issue: For Python 3.12, I've removed the requirement in setup.py for numpy <2, because it's not necessary anymore, because distutils is not used anymore
- Compilation error after the numpy workaround: This is one of the compilation errors with f90wrap that I wanted to draw James's attention to (and as written above, mistakingly created this pull request over).

jameskermode commented:
No worries Svenja. I've convered to a draft PR so it doesn't get merged by accident.

smjanke commented:
Thank you, James!

bernstei commented:
I'll not worry about this for now, and will switch to figuring out if something broken for me in the public branch.

smjanke committed:
Quippy compilation based on Python version enabled

The quippy Makefile now uses different compilation instructions
for Python <3.12 and Python =>3.12, enabling both cases.

smjanke commented:
Small update: I modified the quippy Makefile such that it has separate compilation instructions for Python<3.12 and Python=>3.12.

smjanke committed:
Fix to allow finding lquip_nostub for Py>=3.12

jameskermode commented:
This PR is being closed in favour of #695
